clang/runtime/dpct-rt/include/sparse_utils.hpp

```diff
@@ -743,6 +767,316 @@ inline void spmm(sycl::queue queue, oneapi::mkl::transpose trans_a,
+                            sparse_matrix_desc_t a, sparse_matrix_desc_t b,
+                            const void *beta, sparse_matrix_desc_t c,
+                            oneapi::mkl::sparse::matmat_descr_t matmat_descr) {
+  set_matmat_data(matmat_descr, oneapi::mkl::sparse::matrix_view_descr::general,
+                  trans_a, oneapi::mkl::sparse::matrix_view_descr::general,
+                  trans_b, oneapi::mkl::sparse::matrix_view_descr::general);
+  oneapi::mkl::sparse::matmat(queue, a->get_matrix_handle(),
+                              b->get_matrix_handle(), c->get_matrix_handle(),
+                              oneapi::mkl::sparse::matmat_request::finalize,
+                              matmat_descr, nullptr, nullptr, {});
+  queue.wait();
+}
+
+namespace detail {
+template <typename T>
+inline void check_alpha_value(const void *alpha, sycl::queue queue) {
+  auto alpha_value =
+      dpct::detail::get_value(reinterpret_cast<const T *>(alpha), queue);
+  if (alpha_value != T(1.0f)) {
+    throw std::runtime_error(
+        "oneapi::mkl::sparse::trsv only supports alpha == 1.");
+  }
+}
+} // namespace detail
+
+/// Performs internal optimizations for spsv by analyzing the provided matrix
+/// structure and operation parameters.
+/// \param [in] queue The queue where the routine should be executed. It must
+/// have the in_order property when using the USM mode.
+/// \param [in] trans_a Specifies operation on input matrix a.
+/// \param [in] a Specifies the sparse matrix a.
+inline void spsv_optimize(sycl::queue queue, oneapi::mkl::transpose trans_a,
+                          sparse_matrix_desc_t a) {
+  if (!a->get_uplo() || !a->get_diag()) {
+    throw std::runtime_error("oneapi::mkl::sparse::trsv needs uplo and diag "
+                             "attributes to be specified.");
+  }
+  oneapi::mkl::sparse::optimize_trsv(
+      queue, a->get_uplo().value(), oneapi::mkl::transpose::nontrans,
+      a->get_diag().value(), a->get_matrix_handle());
+}
+
+/// Solves a system of linear equations for a sparse matrix.
+/// \param [in] queue The queue where the routine should be executed. It must
+/// have the in_order property when using the USM mode.
+/// \param [in] trans_a Specifies operation on input matrix a.
+/// \param [in] alpha Specifies the scalar alpha.
+/// \param [in] a Specifies the sparse matrix a.
+/// \param [in] x Specifies the dense vector x.
+/// \param [in, out] y Specifies the dense vector y.
+/// \param [in] data_type Specifies the data type of \param a, \param x and
+/// \param y .
+inline void spsv(sycl::queue queue, oneapi::mkl::transpose trans_a,
+                 const void *alpha, sparse_matrix_desc_t a,
+                 std::shared_ptr<dense_vector_desc> x,
+                 std::shared_ptr<dense_vector_desc> y,
+                 library_data_t data_type) {
+  switch (data_type) {
+  case library_data_t::real_float:
+    detail::check_alpha_value<float>(alpha, queue);
```

Comment:
Does this check include checking if `alpha` is on host or on device? oneMKL Sparse BLAS does not support `alpha` being on device. `alpha` is passed by value on the host to `sparse::trsv()`. Are you taking care of this?

Reply:
Yes. We checked it. If it is device memory, we will copy it to host and then do comparing.